src/rust/engine/fs/src/store.rs

```diff
@@ -256,54 +274,200 @@ mod local {
+            );
+            self.aged_fingerprints(
+              &txn,
+              self.inner.directory_database,
+              &mut used_bytes,
+              &mut fingerprints_by_expired_ago,
+            );
+          }
+          while used_bytes > target_bytes {
+            let aged_fingerprint = fingerprints_by_expired_ago.pop().expect(
+              "lmdb corruption detected, sum of size of blobs exceeded stored blobs",
+            );
+            if aged_fingerprint.expired_seconds_ago == 0 {
+              // Ran out of expired blobs - everything remaining is leased and cannot be collected.
+              return Err(MapFull);
+            }
+            txn
+              .del(
+                aged_fingerprint.database,
+                &aged_fingerprint.fingerprint.as_ref(),
+                None,
+              )
+              .expect("Failed to delete lmdb blob");
+            used_bytes -= aged_fingerprint.size_bytes;
+          }
+          txn.commit()
+        })
+        .or_else(|e| match e {
+          MapFull => Ok(()),
+          e => Err(format!("Error shrinking store: {:?}", e)),
+        })?;
+      Ok(used_bytes)
+    }
+
+    fn aged_fingerprints<T>(
+      &self,
+      txn: &T,
+      database: Database,
+      used_bytes: &mut usize,
+      fingerprints_by_expired_ago: &mut BinaryHeap<AgedFingerprint>,
+    ) where
+      T: Transaction,
+    {
+      let mut cursor = txn.open_ro_cursor(database).expect(
+        "Failed to open lmdb read cursor",
+      );
+      for (key, bytes) in cursor.iter() {
+        *used_bytes = *used_bytes + bytes.len();
+
+        // Random access into the lease_database is slower than iterating, but hopefully garbage
+        // collection is rare enough that we can get away with this, rather than do two passes here
+        // (either to populate leases into pre-populated AgedFingerprints, or to read sizes when
+        // we delete from lmdb to track how much we've freed).
+        let lease_until_unix_timestamp = txn
+          .get(self.inner.lease_database, &key)
+          .map(|b| LittleEndian::read_u64(b))
+          .unwrap_or_else(|e| match e {
+            NotFound => 0,
+            e => panic!("Error reading lease, probable lmdb corruption: {:?}", e),
+          });
+
+        let leased_until = time::UNIX_EPOCH + time::Duration::from_secs(lease_until_unix_timestamp);
+
+        let expired_seconds_ago = time::SystemTime::now()
+          .duration_since(leased_until)
+          .map(|t| t.as_secs())
+          // 0 indicates unleased.
+          .unwrap_or(0);
+
+        fingerprints_by_expired_ago.push(AgedFingerprint {
+          fingerprint: Fingerprint::from_bytes_unsafe(key),
+          size_bytes: bytes.len(),
+          database: database,
+          expired_seconds_ago: expired_seconds_ago,
+        });
+      }
+    }
+
+    pub fn store_bytes(
+      &self,
+      entry_type: EntryType,
+      bytes: Vec<u8>,
+      initial_lease: bool,
```

Comment:
It's unclear when this should be set... a better name, or a docstring somewhere would be good. AFAICT it's set to true in all cases except in tests?